File: Leetcode/Algorithm/java/Binary Tree Inorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        res = new ArrayList<Integer>();
        do_inorderTraversal(root);
        return res;
    }
    
    private void do_inorderTraversal(TreeNode root) {
        if (root == null) {
            return;
        }
        if (root.left != null) {
            do_inorderTraversal(root.left);
        }
        res.add(root.val);
        if (root.right != null) {
            do_inorderTraversal(root.right);
        }
    }
    private List<Integer> res;
}

File: Leetcode/Algorithm/cpp/Binary Tree Inorder Traversal.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> res;
        stack<TreeNode*> st;
        if (!root) {
            return res;
        }
        TreeNode *p = root;
        while (true) {
            if (p) {
                st.push(p);
                p = p -> left;
            } else if (!st.empty()){
                p = st.top();
                st.pop();
                res.push_back(p -> val);
                p = p -> right;
            } else {
                break;
            }
        }
        return res;
    }
};
